include/zephyr/devicetree.h

```diff
@@ -3017,6 +3017,58 @@
 #define DT_NODE_HAS_STATUS(node_id, status) \
 	DT_NODE_HAS_STATUS_INTERNAL(node_id, status)
 
+/**
+ * @brief Get compatible of a node by index.
+ *
+ * The compatible returned can be used directly in other DT macros that requires
+ * `compat` as argument
+ *
+ * Example devicetree fragment:
+ *
+ * @code{.dts}
+ *     n: node {
+ *             compatible = "vnd,specific-device", "generic-device";
+ *     }
+ * @endcode
+ *
+ * Example usage:
+ *
+ * @code{.c}
+ *     DT_COMPAT_BY_IDX(DT_NODELABEL(n), 0) // vnd_specific_device
+ *     DT_COMPAT_BY_IDX(DT_NODELABEL(n), 1) // generic_device
+ * @endcode
+ *
+ * @param node_id node identifier
+ * @param idx the index to get
+ * @return the compat at idx
+ */
+#define DT_COMPAT_BY_IDX(node_id, idx) DT_STRING_TOKEN_BY_IDX(node_id, compatible, idx)
```

Comment:
If I recall correct, there's no guarantee of the indices meaning:
 ```
 DT_COMPAT_BY_IDX(DT_NODELABEL(n), 0) // vnd_specific_device
 DT_COMPAT_BY_IDX(DT_NODELABEL(n), 1) // generic_device
```
AND
```
 DT_COMPAT_BY_IDX(DT_NODELABEL(n), 0) // generic_device
 DT_COMPAT_BY_IDX(DT_NODELABEL(n), 1) // vnd_specific_device
```

Are both valid by DeviceTree semantics. I'm afraid that this would be incorrectly leveraged to produce potentially inconsistent binaries with the same underlying C source.

Would another reviewer please confirm this concern?